src/pixelmapFeature.js

```diff
@@ -0,0 +1,451 @@
+          result = {
+            index: [idx],
+            found: [m_this.data()[idx]]
+          };
+          return result;
+        }
+      }
+    }
+    return {index: [], found: []};
+  };
+
+  ////////////////////////////////////////////////////////////////////////////
+  /**
+   * Build
+   */
+  ////////////////////////////////////////////////////////////////////////////
+  this._build = function () {
+    /* Set the build time at the start of the call.  A build can result in
+     * drawing a quad, which can trigger a full layer update, which in tern
+     * checks if this feature is built.  Setting the build time avoid calling
+     * this a second time. */
+    m_this.buildTime().modified();
+    if (!m_srcImage) {
+      var src = this.style.get('url')();
+      if (src instanceof Image && src.complete && src.naturalWidth && src.naturalHeight) {
+        /* we have an already laoded image, so we can just use it. */
+        m_srcImage = src;
+        this._computePixelmap();
+      } else if (src) {
+        var defer = new $.Deferred(), prev_onload, prev_onerror;
+        if (src instanceof Image) {
+          /* we have an unloaded image.  Hook to the load and error callbacks
+           * so that when it is loaded we can use it. */
+          m_srcImage = src;
+          prev_onload = src.onload;
+          prev_onerror = src.onerror;
+        } else {
+          /* we were given a url, so construct a new image */
+          m_srcImage = new Image();
+          // Only set the crossOrigin parameter if this is going across origins.
+          if (src.indexOf(':') >= 0 &&
+              src.indexOf('/') === src.indexOf(':') + 1) {
+            m_srcImage.crossOrigin = this.style.get('crossDomain')() || 'anonymous';
+          }
+        }
+        m_srcImage.onload = function () {
+          if (prev_onload) {
+            prev_onload.apply(this, arguments);
+          }
+          /* Only use this image if our pixelmap hasn't changed since we
+           * attached our handler */
+          if (m_this.style.get('url')() === src) {
+            m_info = undefined;
+            m_this._computePixelmap();
+          }
+          defer.resolve();
+        };
+        m_srcImage.onerror = function () {
+          if (prev_onerror) {
+            prev_onerror.apply(this, arguments);
+          }
+          defer.reject();
+        };
+        defer.promise(this);
+        this.layer().addPromise(this);
+        if (!(src instanceof Image)) {
+          m_srcImage.src = src;
+        }
+      }
+    } else if (m_info) {
+      this._computePixelmap();
+    }
+    return m_this;
+  };
+
+  /**
+   * Compute information for this pixelmap image.  It is wasterful to call this
+   * if the pixelmap has already been prepared (it is invalidated by a change
+   * in the image).
+   */
+  this._preparePixelmap = function () {
+    var i, idx, pixelData;
+
+    if (!m_srcImage || !m_srcImage.complete || !m_srcImage.naturalWidth ||
```

Comment:
since this is a repeating code,  may be  in this or next PR we could write a function instead?